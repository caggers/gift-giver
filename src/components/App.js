import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import Gift from './Gift';
import { max_number } from '../helpers';

class App extends Component {
  constructor() {
    super();
    this.state = {
      gifts: [],
    };
  }

  addGift = () => {
    const { gifts } = this.state;
    const ids = this.state.gifts.map(gift => gift.id);

    gifts.push({ id: max_number(ids) + 1 });

    this.setState({ gifts });
  };

  removeGift = id => {
    const gifts = this.state.gifts.filter(gift => gift.id !== id);
    this.setState({ gifts });
  }

  render() {
    return (
      <div>
        <h2>Welcome</h2>
        <h3>To the Gift Giver</h3>
        <div className="gift-list">
          {this.state.gifts.map(gift => {
            return ( 
              <Gift 
                key={gift.id}
                gift={gift}
                removeGift={this.removeGift} 
              />
            )
          })}
        </div>
        <Button className="btn-add" onClick={this.addGift}>
          Add Gift
        </Button>
      </div>
    );
  }
}

export default App;
