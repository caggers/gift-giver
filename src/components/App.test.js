import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

import { render } from 'react-testing-library';

// Basic test for checking the h2
it('renders without crashing', () => {
  const wrapper = shallow(<App />);
  const welcome = <h2>Welcome</h2>;
  expect(wrapper.contains(welcome)).toEqual(true);
});

// Using jest-enzyme sugar
it('renders without crashing', () => {
  const wrapper = shallow(<App />);
  const welcome = <h2>Welcome</h2>;
  expect(wrapper).toContainReact(welcome);
});

// Test using react-testing-library and jest-dom sugar
it('renders Welcome message', () => {
  const { getByText } = render(<App />);
  expect(getByText('Welcome')).toBeInTheDocument();
});

//Basic Testing from Udemy course
describe('App', () => {
  const app = shallow(<App />);
  it('renders correctly', () => {
    expect(app).toMatchSnapshot();
  });

  it('initialises the `state` with an empty list of gifts', () => {
    expect(app.state().gifts).toEqual([]);
  });

  describe('when clicking the `add-gift` button', () => {
    const id = 1;

    beforeEach(() => {
      app.find('.btn-add').simulate('click');
    });

    afterEach(() => {
      app.setState({ gifts: [] });
    });

    it('adds a new gift to `state`', () => {
      expect(app.state().gifts).toEqual([{ id }]);
    });

    it('adds a new gift to the rendered list', () => {
      expect(app.find('.gift-list').children().length).toEqual(1);
    });

    it('creates a Gift component', () => {
      expect(app.find('Gift').exists()).toBe(true);
    });

    describe('and the user wants to remove the added gift', () => {
      beforeEach(() => {
        app.instance().removeGift(id);
      });

      it('removes the gift from `state`', () => {
        expect(app.state().gifts).toEqual([]);
      });
    });
  });
});
