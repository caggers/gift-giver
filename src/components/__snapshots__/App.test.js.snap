// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <h2>
    Welcome
  </h2>
  <h3>
    To the Gift Giver
  </h3>
  <div
    className="gift-list"
  />
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="default"
    className="btn-add"
    disabled={false}
    onClick={[Function]}
  >
    Add Gift
  </Button>
</div>
`;
